File: Python/Contains Duplicate II.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(n)
    """

    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        index = {}

        for i, num in enumerate(nums):
            if num in index and i - index[num] <= k:
                return True
            index[num] = i

        return False

File: Java/Contains Duplicate II.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    /**
     * Time:   O(n)
     * Memory: O(n)
     */
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; ++i) {
            if (map.containsKey(nums[i]) && i - map.get(nums[i]) <= k)
                return true;
            map.put(nums[i], i);
        }

        return false;
    }
}
